server/keyspace_service.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"path"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pingcap/kvproto/pkg/keyspacepb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/tikv/pd/server/keyspace"
+	"github.com/tikv/pd/server/storage/endpoint"
+	"go.etcd.io/etcd/clientv3"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// KeyspaceServer wraps GrpcServer to provide keyspace service.
+type KeyspaceServer struct {
+	*GrpcServer
+}
+
+// getErrorHeader returns corresponding ResponseHeader based on err.
+func (s *KeyspaceServer) getErrorHeader(err error) *pdpb.ResponseHeader {
+	switch err {
+	case keyspace.ErrKeyspaceExists:
+		return s.wrapErrorToHeader(pdpb.ErrorType_DUPLICATED_ENTRY, err.Error())
+	case keyspace.ErrKeyspaceNotFound:
+		return s.wrapErrorToHeader(pdpb.ErrorType_ENTRY_NOT_FOUND, err.Error())
+	default:
+		return s.wrapErrorToHeader(pdpb.ErrorType_UNKNOWN, err.Error())
+	}
+}
+
+// validateRequest checks that server is serving and cluster id matches the requested.
+func (s *KeyspaceServer) validateRequest(header *pdpb.RequestHeader) error {
+	if s.IsClosed() {
```

Comment:
Do we need to check the leader?

Reply:
I don't think it's necessary, since both requests are read-only

Reply:
But the raftcluster will be nil?

Reply:
Ok, I think you are right.